### PR TITLE
Add SafeERC20.forceApprove()

### PR DESCRIPTION
**Pull request by Amxx**
Replaces #3851, and targets master
Fixes LIB-705

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Changeset entry (run `npx changeset add`)

### CODE REVIEW

File: contracts/token/ERC20/utils/SafeERC20.sol [hunk 1]
```diff
@@ -47,18 +47,28 @@ library SafeERC20 {
 
     function safeIncreaseAllowance(IERC20 token, address spender, uint256 value) internal {
         uint256 newAllowance = token.allowance(address(this), spender) + value;
-        _callOptionalReturn(token, abi.encodeWithSelector(token.approve.selector, spender, newAllowance));
+        forceApprove(token, spender, newAllowance);
```
Comment (frangio) [inline contracts/token/ERC20/utils/SafeERC20.sol:50]: Hm, I would avoid making this change in 4.9.

File: contracts/token/ERC20/utils/SafeERC20.sol [hunk 2]
```diff
@@ -47,18 +47,28 @@ library SafeERC20 {
...
+            forceApprove(token, spender, newAllowance);
         }
     }
 
+    function forceApprove(IERC20 token, address spender, uint256 value) internal {
+        // solidity has lazy evaluation
+        require(
+            _callOptionalReturnBool(token, abi.encodeWithSelector(token.approve.selector, spender, value)) ||
+                (_callOptionalReturnBool(token, abi.encodeWithSelector(token.approve.selector, spender, 0)) &&
+                    _callOptionalReturnBool(token, abi.encodeWithSelector(token.approve.selector, spender, value))),
```
Comment (frangio) [inline contracts/token/ERC20/utils/SafeERC20.sol:67]: We shuld be able to reuse the encoded calldata in this second call.

File: contracts/token/ERC20/utils/SafeERC20.sol [hunk 3]
```diff
@@ -47,18 +47,28 @@ library SafeERC20 {
...
             uint256 oldAllowance = token.allowance(address(this), spender);
             require(oldAllowance >= value, "SafeERC20: decreased allowance below zero");
             uint256 newAllowance = oldAllowance - value;
-            _callOptionalReturn(token, abi.encodeWithSelector(token.approve.selector, spender, newAllowance));
+            forceApprove(token, spender, newAllowance);
         }
     }
 
+    function forceApprove(IERC20 token, address spender, uint256 value) internal {
+        // solidity has lazy evaluation
```
Comment (frangio) [inline contracts/token/ERC20/utils/SafeERC20.sol:63]: "Short-circuit evaluation" is the right term here.

I think this deserves more comments explaining what's going on.

File: contracts/token/ERC20/utils/SafeERC20.sol [hunk 4]
```diff
@@ -47,18 +47,28 @@ library SafeERC20 {
...
         unchecked {
             uint256 oldAllowance = token.allowance(address(this), spender);
             require(oldAllowance >= value, "SafeERC20: decreased allowance below zero");
             uint256 newAllowance = oldAllowance - value;
-            _callOptionalReturn(token, abi.encodeWithSelector(token.approve.selector, spender, newAllowance));
+            forceApprove(token, spender, newAllowance);
         }
     }
 
+    function forceApprove(IERC20 token, address spender, uint256 value) internal {
```
Comment (frangio) [inline contracts/token/ERC20/utils/SafeERC20.sol:62]: Missing docs.

I realize a lot of this contract is missing docs... If we can add it in this PR I think that would be a great side effect.

File: contracts/token/ERC20/utils/SafeERC20.sol (hunk 2 above)
Comment (frangio) [inline contracts/token/ERC20/utils/SafeERC20.sol:67]: In these 2 calls, why are we not bubbling reverts?

It feels like we should just do 2 normal `_callOptionalReturn`, separated by `;`.

File: test/token/ERC20/utils/SafeERC20.test.js [hunk 5]
```diff
@@ -165,6 +166,43 @@ contract('SafeERC20', function (accounts) {
...
+          this.mock.$safeApprove(this.token.address, spender, 200),
+          'SafeERC20: approve from non-zero to non-zero allowance',
+        );
+      });
+
+      it('safeApproval can update approval to zero', async function () {
+        await this.mock.$safeApprove(this.token.address, spender, 0);
+      });
+
+      it('safeApproval can increasse approval', async function () {
```
Comment (frangio) [inline test/token/ERC20/utils/SafeERC20.test.js:193]: I don't know why this isn't caught by codespell.

```suggestion
      it('safeApproval can increase approval', async function () {
```

File: test/token/ERC20/utils/SafeERC20.test.js [hunk 6]
```diff
@@ -165,6 +166,43 @@ contract('SafeERC20', function (accounts) {
...
+        await this.mock.$safeIncreaseAllowance(this.token.address, spender, 10);
+      });
+
+      it('safeApproval can decreasse approval', async function () {
```
Comment (frangio) [inline test/token/ERC20/utils/SafeERC20.test.js:197]:
```suggestion
      it('safeApproval can decrease approval', async function () {
```

File: test/token/ERC20/utils/SafeERC20.test.js [hunk 7]
```diff
@@ -192,8 +230,17 @@ function shouldRevertOnAllCalls(reason) {
     // [TODO] make sure it's reverting for the right reason
     await expectRevert.unspecified(this.mock.$safeDecreaseAllowance(this.token.address, constants.ZERO_ADDRESS, 0));
   });
+
+  it('reverts on forceApprove', async function () {
+    // [TODO] make sure it's reverting for the right reason
```
Comment (frangio) [inline test/token/ERC20/utils/SafeERC20.test.js:235]: Do you plan to address this TODO in this PR?

File: test/token/ERC20/utils/SafeERC20.test.js [hunk 8]
```diff
@@ -165,6 +166,43 @@ contract('SafeERC20', function (accounts) {
...
+
+      it('safeApproval can increasse approval', async function () {
+        await this.mock.$safeIncreaseAllowance(this.token.address, spender, 10);
+      });
+
+      it('safeApproval can decreasse approval', async function () {
+        await this.mock.$safeDecreaseAllowance(this.token.address, spender, 10);
+      });
+
+      it('forceApproval works', async function () {
```
Comment (frangio) [inline test/token/ERC20/utils/SafeERC20.test.js:201]: We are using "forceApproval" and "safeApproval" in the test names, but the functions are called "forceApprove", etc.

File: contracts/token/ERC20/utils/SafeERC20.sol (hunk 4 above)
Comment (Amxx) [reply contracts/token/ERC20/utils/SafeERC20.sol:62]: Yes, I'll also improve the test of the whole library

File: test/token/ERC20/utils/SafeERC20.test.js (hunk 7 above)
Comment (Amxx) [reply test/token/ERC20/utils/SafeERC20.test.js:235]: This TODO was copied from the other TODO in this file. 

I'll rework the test file completelly

**Review by frangio — COMMENTED** [review]
Just a comment, rest is good!

**Review by frangio — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**